### PR TITLE
Minor fixes around table scan predicate push down

### PR DESCRIPTION
**Pull request by kokosing**
This relates to discussion in https://github.com/prestodb/presto/issues/11134

### CODE REVIEW

**Comment by haozhun** [comment]
I think "Merge table scan original constraint with predicate" is problematic. If you have a filter on top of a TableScan that cannot be pushed down. With your new code, the filter would be repeatedly added to the `originalConstraint` field of `TableScanNode`. This goes against the intention of `originalConstraint`, which needs to be succinct for human reading purpose.

On the other hand, I can also see why doing an "and" here is necessary. The current way of handling `originalConstraint` could be producing a misleading plan. I also have question about the usefulness of it, if we move to make TPCH and Hive connector implement this field properly.

**Comment by kokosing** [comment]
> I also have question about the usefulness of it, if we move to make TPCH and Hive connector implement this field properly.

Even if connector supports predicate pushdown fully a table may have high number of partitions, converting tuple domain to predicate will result in huge `IN` expression. However, we could simplify tuple domain before converting this predicate and then `IN` should be replaced with range predicates.

Will remove problematic commit out of this PR.

**Comment by kokosing** [comment]
>  However, we could simplify tuple domain before converting this predicate and then IN should be replaced with range predicates.

It is already done this way.

**Comment by haozhun** [comment]
### "Remove unused field"

looks good

### "Return proper layout predicate for cassandra"

Assuming it's correct and doesn't lead to other problems (like over-complicating the predicate), this would be something nice to have. But this isn't necessary (and maybe not useful either). My initial assessment in #11134 on what `currentConstraint` should be is wrong, and thus some conclusions and proposals that came after that. Quote what I put down in javadoc for `getCurrentConstraint` in #11322,

```
    * A TupleDomain that represents a predicate that every row this TableScan node
     * produces is guaranteed to satisfy.
     * <p>
     * This guarantee can have different origins.
     * For example, it may be successful predicate push down, or inherent guarantee provided by the underlying data.
```

Several issues contributed to this misunderstanding:
* No documentation on `getCurrentConstraint`
* A long-standing bug (use intersection of `newDomain` and `ConnectorTableLayout.predicate` as `currentConstraint`)
* Allow both `PickTableLayout` and `AddExchange` to happen, and make `PickTableLayout` iterative. (This change depends on the bug.)

### "Rename variable"

Similar problems will unavoidably happen again if we continue to have two copies of this code. That's why I merged them in #11322. This commit will no longer be applicable after that.

### "Use layout predicate as table scan current constraint"

Removing the intersection is not ok. The semantics of currentConstraint does not guarantee accuracy.

This test fixes is likely going to be useful (and applicable) to my changes in #11322. Thank you!

### "Use Optional for TableScanNode originalConstraint"

With your #11175, this is no longer applicable.

**Comment by kokosing** [comment]
> Assuming it's correct and doesn't lead to other problems (like over-complicating the predicate)

In the worst case it is not worse than what we can have in hive I think.

> The semantics of currentConstraint does not guarantee accuracy. Future invocation of Metadata.getTableLayouts could lose filter as a result.

I think I might be missing something. When `Metadata.getTableLayouts` is called several times we need to pass the proper predicate (constraint), so why we can lose the filter result? We cannot lose the accuracy because it could lead to wrong query results.

**Comment by haozhun** [comment]
As I said in my earlier comment, I'm totally fine with the Cassandra changes. It's up to you.

---------------

"We cannot lose the accuracy because it could lead to wrong query results." - It is true that losing accuracy would lead to wrong query results.

"When Metadata.getTableLayouts is called several times we need to pass the proper predicate (constraint)" - It is true that you **need** to.

Prior to this PR, `currentConstraint` is de facto guaranteed to be accurate because it is set to `newDomain.intersect(layout.getLayout().getPredicate())`. However, this is not a guarantee provided by the intended semantics of `currentConstraint` (A TupleDomain that represents a predicate that every row this TableScan node produces is guaranteed to satisfy).

By removing this `intersect`, this de facto guarantee no longer holds. The result could be incorrect if one invokes getTableLayouts repeatedly (using currentConstraint in subsequent calls). Addressing this issue is the whole point of #11322. Presumably, this is also why you had to change Cassandra connector before you can remove this intersect.

**Comment by kokosing** [comment]
> By removing this intersect, this de facto guarantee no longer holds. The result could be incorrect if one invokes getTableLayouts repeatedly (using currentConstraint in subsequent calls). Addressing this issue is the whole point of #11322. Presumably, this is also why you had to change Cassandra connector before you can remove this intersect.

Among information that connector returns for `Metadata.getTableLayouts` is a predicate that connector was able to push down `ConnectorTableLayout#predicate` and what it was not able `ConnectorTableLayoutResult#unenforcedConstraint`. Here cassandra was correctly returning only the latter, and `TupleDomai.all()` for former. I would love to have a connector contract that these both things are coherent and returned properly according to connector capabilities. As I understand in #11322 you only require `ConnectorTableLayoutResult#unenforcedConstraint` to be correct.

I think we could add validation for `Metadata.getTableLayouts` that union of `ConnectorTableLayout#predicate` and `ConnectorTableLayoutResult#unenforcedConstraint` is completely enclosed by predicate `TupleDomain`. That could catch the cassandra error.

**Comment by kokosing** [comment]
> I think we could add validation for Metadata.getTableLayouts that union of ...

Sorry I meant `intersection`.

I applied your comments.

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java [hunk 1]
```diff
@@ -351,11 +352,19 @@ public boolean catalogExists(Session session, String catalogName)
         ConnectorSession connectorSession = session.toConnectorSession(connectorId);
         List<ConnectorTableLayoutResult> layouts = metadata.getTableLayouts(connectorSession, connectorTable, constraint, desiredColumns);
 
+        layouts.stream()
+                .forEach(layout -> checkState(layoutMatchesConstraint(constraint, layout), "Returned layout does not match the constraint"));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java]: This was catching the problem in the cassandra fixed in the previous commit.

**Comment by haozhun** [comment]
> Among information that connector returns for Metadata.getTableLayouts is a predicate that connector was able to push down ConnectorTableLayout#predicate

This is not correct. As I clarified in #11322, `predicate` stands for a predicate that every row the connector provides satisfies. It may not even be a predicate that was pushed down (i.e. it can be some inherent guarantee provided by the data. see order status in tpch connector).

> and what it was not able ConnectorTableLayoutResult#unenforcedConstraint. 

Yes.

> Here cassandra was correctly returning only the latter, and TupleDomai.all() for former.

No, this is true because the former because `true` is a predicate that every row the connector provides satisfies.

> I would love to have a connector contract that these both things are coherent and returned properly according to connector capabilities. As I understand in #11322 you only require ConnectorTableLayoutResult#unenforcedConstraint to be correct.

This argument does not apply due to reasons stated above.

**Pull request closed without merging**